### ISSUE TITLE
14 Bugs and feature requests at once

### ISSUE THREAD

**Issue by Dragon8256** [ViaVersion/ViaFabricPlus#178]
### General Information

1.20.1 ViaFabricPlus latest build (2.8.0-SNAPSHOT) Fabric api 0.85.0 Fabric loader 0.14.21
If i join from direct version i use betacraft launcher
(Sorry for bad english anywhere here)

### Bug Description

1. On alpha versions and beta to b1.2 blocks may not break. Sending more mining packets may solve this because when i enable fly with hack client (this slow down the breaking) and speed up the game blocks break without problems. Also related to this some blocks in old versions (crafting table, furnace, chest) have longer breaking time and i can't break this blocks without slowing down game to about 0.1x (Recorded in a1.1.2_02)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/69464ec7-753d-40f8-b54c-1583b659461d

2. I can use swords to block before b1.8 (Recorded in a1.1.2_02)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/3910c84b-56fd-4ae0-a7fc-b956b9335f1a

3. Pasting items from saved hotbars work weird in ~1.12.2 and lower. This caused by changed nbt tags in 1.13. Item names turn to something like {"text":"name"} enchanted golden apples don't paste at all because in 1.12.2 enchanted golden apple is variation of default golden apple, but in 1.13 this is separate item, cobwebs also don't paste. Enchantments can't be pasted even with enchantment books. Also will be good to implement auto convert of netherite items to diamond (Recorded in 1.12.2)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/7e37958f-f163-4f07-a9a8-dab75ce22973

4. Converting /setblock and /fill commands with block data will be good for pasting schematics with litematica (Recorded in 1.12.2)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/3a82bb0a-b246-401e-ace8-20fbc4a66f52

5. Pistons don't push (Recorded in 1.12.2)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/327eebe7-2ecf-4b11-b17a-dce78e5e431e

6. Block transmutation can create blocks that wasn't added in this version. Not sure how this can be fixed and if it should be fixed at all (recorded in b1.8.1)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/a9241b91-8453-4994-bf15-fde9e67e08a7
https://youtu.be/MF-KR6A3KyM

7. After respawn skin not show. The skin is returned after rejoin on server (Recorded in b1.8.1, also noticed this bug in a1.1.2)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/c6238e1d-9bb7-4ee4-9b4e-36ed7604049b

8. When i place redstone dust going upwards it don't connect to lower redstone dust (Recorded in b1.8.1)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/3243d1db-3a53-4e6b-a837-976bbd98d1ae

9. Piston heads has weird hitbox (Recorded in b1.8.1)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/2851704a-e33f-40b8-868a-edd572333183

10. If online-mode=true i can't connect to versions a1.1.2-a1.2.6 (at least this versions, maybe there are more versions with this problem)

11. As alpha has non-verifiable inventory, when i place blocks in protected world spawn area they just disappear

12. Spiders don't make knockback, other mobs make (Recorded in a1.1.2_02)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/8cf03587-39ad-4482-aa30-a94c4437c9e0

13. Swords can break blocks in creative in b1.8 and in some versions after that. This was in the original b1.8 but would be good if it gets fixed

14. Shift+right click on block that has states (e.g. repeater) will cause repeater to increase delay by 1. Same situation as 13th

### Steps to Reproduce

I think i explained everything above, if you still have questions how to reproduce this bugs i will answer 

### Expected Behavior

1. Blocks should break normally
2. Swords should not block
5. Pistons should push player
7. Skin should show after respawn
8. Redstone should work fine
9. Piston hitbox should be as normal
10. I should connect to this versions with online-mode=true
11. Blocks should be not placed at all/returned after not successful place
12. Spiders should knockback

**Comment by RaphiMC** [comment]
Hey, thanks for your very extensive bug reports and feature requests. Due to ViaFabricPlus itself not containing any protocol translation, but only the clientside fixes, many of those bugs have to get addresses in the underlying libraries.

1. Could maybe be a ViaLegacy bug (Not sure if due to how VFP sends player position packets it might be broken, but blocks like grass didn't ever have their block mining time changed afaik. The other blocks you mentioned have wrong timing in VFP)
2. VFP bug most likely, because swords can't even be blocked regularly in 1.9+ (I could maybe try to fix it in ViaLegacy as well by forcing the client to unblock)
3. ViaVersion bug (ViaVersion intentionally left shulker boxes broken in <= 1.12.2 due to performance/security concerns)
4. Good feature to get implemented in VFP
5. VFP bug
6. ViaLegacy probably
7. ViaLegacy probably
8. ViaVersion bug probably
9. VFP bug
10. Online mode doesn't work on a vanilla client either in those versions. Mojang shut down that API
11. VFP didn't implement the AlphaInventoryProvider from ViaLegacy. It could make the clientside inventory a lot more close to reality by using the actual clientside inventory instead of relying on tracked data
12. Maybe a ViaLegacy bug
13. I don't agree that this should be fixed as its intentional behaviour. @FlorianMichael Might maybe could add an option to disable it
14. You didn't mention a server version so idk

To properly verify some of the non VFP bugs I want you to replicate them using ViaProxy and a vanilla 1.20.1 client without mods. Additionally verify if whatever you test actually is broken or if the vanilla client has the same behaviour (12 maybe)

**Comment by Dragon8256** [comment]
Not sure i should write this here or as separate issue to ViaProxy but will write here.
1. This is very random problem, when i tested first time i broke ~60 blocks and each broke 1st try. But then this bug returned even though I didn't change anything. Found new very strange bug, with ViaProxy i can build inside spawn-protected region in a1.1.2_01 (server restart don't fix this and this was new world)
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/d5d26e5c-1251-44ed-bc39-c5746bbca940

3. I think will be good to convert at least enchantments, because without this only option is /enchant

6. I think in VFP should be option to change blocks to how them look in version you connect. in b1.8.1 look like this:
![image](https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/6ec65fe0-674d-4bed-904d-15b06a61f6b4)

7. Skin still disappear with ViaProxy

8. Redstone still broken with ViaProxy, i think solution is to simulate block update to redstone on client/proxy

9. In ViaProxy piston head has correct placement hitbox but wrong collision hitbox, anticheat just tp me back

12. This is vanilla behavior and not need to be fixed (also this is a1.2.6, not a1.1.2)

14. This was also b1.8.1, beta don't have function to shift + right click and place block instead of interact with block. And yes this is intentional behavior, so same situation as 13th

**Comment by RaphiMC** [comment]
I fixed the ViaLegacy issues, the remaining issues have to be fixed in VFP

**Comment by FlorianMichael** [comment]
Hello! 
The issues 2, 5, 9 and 11 should be fixed now, I won't fix issue 13/14 because that's vanilla behavior unless you have a good reason for it. For Issue 3 I would ask you to make an issue on the main repository ViaVersion. Issue 4 is now in my internal TODO list and it will be done if I have time for it.

You can download the latest development release here and test if everything works like this: https://github.com/ViaVersion/ViaFabricPlus/actions/runs/5583206503

**Comment by Dragon8256** [comment]
Everything works fine except for 11. Repeated what i wrote in previous comment - i can build inside spawn protected region, while correct version players can't (Tested in a1.1.2 and a1.2.6, same result)

https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/5d05f10c-f1ab-4688-a6bb-f98fddb8e2c7
[Server a1.1.2.zip](https://github.com/ViaVersion/ViaFabricPlus/files/12076666/Server.a1.1.2.zip)

**Comment by FlorianMichael** [comment]
I can't reproduce that, could it be, that one of the accounts maybe has operator permissions?

**Comment by Dragon8256** [comment]
Yes, my account had operator permissions. I removed my account from ops.txt now result is this
https://github.com/ViaVersion/ViaFabricPlus/assets/123386522/81745d3d-e804-481a-9f05-4c9948cc6f75

Maybe possible to make ghost blocks not appear at all and return blocks to inventory

**Comment by FlorianMichael** [comment]
You can see that the same bug happens in the normal Alpha Client as well, so it's not a ViaFabricPlus bug.

**Comment by FlorianMichael** [comment]
I will then close the issue, since the AlphaClient shows the same behavior as ViaFabricPlus, should there still be problems, or should I be wrong, then you can open the issue again.
